packages/sit-onyx/src/components/TestInput/TestInput.vue

```diff
@@ -1,27 +1,59 @@
 <script lang="ts" setup>
-import { computed } from "vue";
-
-const props = withDefaults(
-  defineProps<{
-    /** The current input value */
-    modelValue?: string;
-    /** Label to show next to the input */
-    label?: string;
-    required?: boolean;
-  }>(),
-  {
-    modelValue: "",
-    label: "",
-  },
-);
+import { computed, onMounted, ref, watch } from "vue";
+
+type NativeInputProps = Partial<{
+  required: boolean;
+  pattern: string;
+  // TODO: we probably want to have number inputs separated from the others in the future.
+  type: "email" | "number" | "password" | "search" | "tel" | "text" | "url";
+  max: number | string;
+  maxLength: number;
+  min: number | string;
+  minLength: number;
+  title: string;
+}>;
+export type TestInputProps = {
+  /** The current input value
+   * TODO: remove the "number" once we separated number inputs from the other types.
+   */
+  modelValue?: string | number;
+  /** Label to show next to the input */
+  label?: string;
+  /** A custom error message that will be shown (only when the input is invalid) */
+  customErrorMessage?: string;
+} & NativeInputProps;
+
+const props = withDefaults(defineProps<TestInputProps>(), {
+  modelValue: "",
+  label: "",
+  type: "text",
+  customErrorMessage: undefined,
+});
 
 const emit = defineEmits<{
-  /** Emitted when input value changes */
-  "update:modelValue": [value: string];
+  /** Emitted when input value changes
+   * TODO: remove the "number" once we separated number inputs from the other types.
+   */
+  "update:modelValue": [value: string | number];
   /** Emitted on blur when the input value changes */
   change: [value: string];
+  /** Emitted whenever the validity state of the input changes */
+  validityChange: [state: ValidityState];
 }>();
 
+const isTouched = ref(false);
+
+const coreElement = ref<HTMLInputElement | null>(null);
+
+const isValid = ref(coreElement.value?.validity.valid || false);
+
+const getErrorMessage = (): string => {
+  if (isValid.value) return "";
+
+  return props.customErrorMessage || coreElement.value?.validationMessage || "";
+};
+const errorMessage = ref(getErrorMessage());
```

Comment:
change isValid to contain the object, then the errorMessage can be a computed again.
```suggestion
const validityState = ref(coreElement.value?.validity || null);

const getErrorMessage = (): string => {
  if (isValid.value) return "";

  return props.customErrorMessage || coreElement.value?.validationMessage || "";
};
const errorMessage = ref(getErrorMessage());
```